WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -65,12 +65,49 @@ public IEnumerable<Money> Decompose(IEnumerable<Coin> myInputCoins, IEnumerable<
 				}
 			}
 
-			if (remaining >= MinimumAmountPlusFee)
+			if (remaining >= MinimumAmountPlusFee && remainingVsize >= OutputSize)
 			{
-				outputAmounts.Add(remaining - OutputFee);
+				naiveSet.Add(remaining - OutputFee);
+			}
+
+			setCandidates.Add(naiveSet.Count, naiveSet);
+			var before = DateTimeOffset.UtcNow;
+			do
+			{
+				remaining = inputs.Sum();
+				remainingVsize = AvailableVsize;
+				var currSet = new List<Money>();
+				do
+				{
+					var denomPlusFees = denoms.Where(x => x <= remaining && x >= (remaining / 3)).ToList();
+					var denomPlusFee = denomPlusFees.RandomElement();
+					if (denomPlusFee is null || remaining < MinimumAmountPlusFee || remainingVsize < 2 * OutputSize)
+					{
+						break;
+					}
+
+					if (denomPlusFee <= remaining)
+					{
+						currSet.Add(denomPlusFee - OutputFee);
+						remaining -= denomPlusFee;
+						remainingVsize -= OutputSize;
+					}
+				}
+				while (currSet.Count <= naiveSet.Count || currSet.Count <= 3);
+
+				if (currSet.Count <= naiveSet.Count || currSet.Count <= 3)
+				{
+					if (remaining >= MinimumAmountPlusFee && remainingVsize >= OutputSize)
+					{
+						currSet.Add(remaining);
+					}
+
+					setCandidates.TryAdd(currSet.Count, currSet);
+				}
 			}
+			while ((DateTimeOffset.UtcNow - before).TotalMilliseconds <= 30);
```

Comment:
This can be longer. It was this short for me only because in the tests I had to run it hundreds of times as I was simulating hundreds of users.